x/metadata/keeper/signers_utils.go

```diff
@@ -0,0 +1,348 @@
+
+// authzCacheContextKey is the key used in an sdk.Context to set/get the AuthzCache.
+const authzCacheContextKey = "authzCacheContextKey"
+
+// AddAuthzCacheToContext either returns a new sdk.Context with the addition of an AuthzCache,
+// or clears out the AuthzCache if it already exists in the context.
+// It panics if the AuthzCache key exists in the context but isn't an AuthzCache.
+func AddAuthzCacheToContext(ctx sdk.Context) sdk.Context {
+	// If it's already got one, leave it there but clear it out.
+	// Otherwise, we'll add a new one.
+	if cacheV := ctx.Value(authzCacheContextKey); cacheV != nil {
+		if cache, ok := cacheV.(*AuthzCache); ok {
+			cache.Clear()
+			return ctx
+		}
+		// If the key was there, but not an AuthzCache, things are very wrong. Panic.
+		panic(fmt.Errorf("context value %q is a %T, expected %T",
+			authzCacheContextKey, cacheV, NewAuthzCache()))
+	}
+	return ctx.WithValue(authzCacheContextKey, NewAuthzCache())
+}
+
+// GetAuthzCache gets the AuthzCache from the context or panics.
+func GetAuthzCache(ctx sdk.Context) *AuthzCache {
+	cacheV := ctx.Value(authzCacheContextKey)
+	if cacheV == nil {
+		panic(fmt.Errorf("context does not contain a %q value", authzCacheContextKey))
+	}
+	cache, ok := cacheV.(*AuthzCache)
+	if !ok {
+		panic(fmt.Errorf("context value %q is a %T, expected %T",
+			authzCacheContextKey, cacheV, NewAuthzCache()))
+	}
+	return cache
+}
+
+// UnwrapMetadataContext retrieves a Context from a context.Context instance attached with WrapSDKContext.
+// It then adds an AuthzCache to it.
+// It panics if a Context was not properly attached, or if the AuthzCache can't be added.
+//
+// This should be used for all Metadata msg server endpoints instead of sdk.UnwrapSDKContext.
+// This should not be used outside of the Metadata module.
+func UnwrapMetadataContext(goCtx context.Context) sdk.Context {
+	return AddAuthzCacheToContext(sdk.UnwrapSDKContext(goCtx))
+}
+
+// findMissing returns all elements of the required list that are not found in the entries list.
+//
+// See also: findMissingComp
+func findMissing(required, toCheck []string) []string {
+	return findMissingComp(required, toCheck, func(r, c string) bool { return r == c })
+}
+
+// findMissingParties returns all parties from the required list that don't have a same party in the toCheck list.
+//
+// See also: findMissingComp
+func findMissingParties(required, toCheck []types.Party) []types.Party {
+	return findMissingComp(required, toCheck, func(r, c types.Party) bool { return types.SamePartiers(&r, &c) })
+}
+
+// findMissingComp returns all entries in required where an entry does not exist in toCheck
+// such that the provided comp function returns true.
+// Duplicate entries in required do not require duplicate entries in toCheck.
+// E.g. findMissingComp([a, b, a], [a]) => [b], and findMissingComp([a, b, a], [b]) => [a, a].
+func findMissingComp[R any, C any](required []R, toCheck []C, comp func(R, C) bool) []R {
+	var rv []R
+reqLoop:
```

Comment:
Same goto comment as above.

Reply:
With these loops with labels, I didn't have any need for the just the inner loop functionality (e.g. a `containsComp` function). So, rather than create them and add unit tests on them just so they could be used in the one place I was designing them for, I opted for nesting the loops directly.

Also, I tend to prefer the golang loop labels over the `found` bool way.

And in most of these, extracting the inner loop into its own function, and unit testing that function, doesn't really reduce the unit tests needed on the original function. So it's extra work for something that isn't really needed.